views/js/portableElementRegistry/factory/factory.js

```diff
@@ -0,0 +1,200 @@
+                    }
+                }
+            },
+            registerProvider : function registerProvider(moduleName){
+                __providers[moduleName] = null;
+                return this;
+            },
+            resetProviders : function resetProviders(){
+                __providers = {};
+                return this;
+            },
+            loadProviders : function loadProviders(callback){
+                var self = this;
+                var providerLoadingStack = [];
+                _.forIn(__providers, function(provider, id){
+                    if(provider === null){
+                        providerLoadingStack.push(id);
+                    }
+                });
+                _requirejs(providerLoadingStack, function(){
+                    _.each([].slice.call(arguments), function(provider){
+                        if(provider && _.isFunction(provider.load)){
+                            __providers[providerLoadingStack.shift()] = provider;
+                        }
+                    });
+                    callback();
+                    self.trigger('providersloaded');
+                });
+            },
+            getAllVersions : function getAllVersions(){
+                var all = {};
+                _.forIn(this._registry, function (versions, id){
+                    all[id] = _.map(versions, 'version');
+                });
+                return all;
+            },
+            getRuntime : function getRuntime(typeIdentifier, version){
+                var pci = this.get(typeIdentifier, version);
+                if(pci){
+                    return _.assign(pci.runtime, {
+                        id : pci.typeIdentifier,
+                        label : pci.label,
+                        baseUrl : pci.baseUrl
+                    });
+                }else{
+                    this.trigger('error', 'no portable element runtime found' , typeIdentifier, version);
+                }
+            },
+            getCreator : function getCreator(typeIdentifier, version){
+                var pci = this.get(typeIdentifier, version);
+                if(pci && pci.creator){
+                    return _.assign(pci.creator, {
+                        id : pci.typeIdentifier,
+                        label : pci.label,
+                        baseUrl : pci.baseUrl,
+                        response : pci.response
+                    });
+                }else{
+                    this.trigger('error', 'no portable element creator found' , typeIdentifier, version);
+                }
+            },
+            getBaseUrl : function getBaseUrl(typeIdentifier, version){
+                var runtime = this.get(typeIdentifier, version);
+                if(runtime){
+                    return runtime.baseUrl;
+                }
+                return '';
+            },
+            loadRuntimes : function loadRuntimes(callback, reload){
```

Comment:
Why do you use a callback instead of a promise ? The global architecture could be simplified